db/revision_cache_lru.go

```diff
@@ -265,13 +266,13 @@ func (rc *LRURevisionCache) Upsert(ctx context.Context, docRev DocumentRevision,
 	}
 
 	// Purge oldest item if over number capacity
-	var numItemsRemoved int
-	for len(rc.cache) > int(rc.capacity) {
-		rc.purgeOldest_()
-		numItemsRemoved++
+	numItemsRemoved, numBytesEvicted := rc._numberCapacityEviction()
+	if numBytesEvicted > 0 {
+		rc.updateRevCacheMemoryUsage(-numBytesEvicted)
 	}
+	rc.lock.Unlock() // release lock after eviction finished
```

Comment:
rc.updateRevCacheMemoryUsage doesn't require a lock, it updates atomics, so we can move this up.

Reply:
Issue is if we don't update whilst holding lock and another goroutine is waiting on acquiring lock for eviction once this is released the other goroutine newly acquiring lock will be acting on incorrect memory stats and will over evict.

Reply:
It kinda looks like everywhere we decrement this we need to be holding the lock, and everywhere we increment we don't. Is that right? Makes me wonder if they should be separate functions with better indication of which lock is required.